### PR TITLE
[Heapster] Reduce catchpoints and add more Mbox proofs

### PR DESCRIPTION
**Pull request by m-yac**
This PR has two parts:
1. Change how the catch rule in the Heapster implication checker is translated, so that branches that could fail are pruned in favor of branches that are guaranteed not to, and update the generated files and proofs in the heapster examples directory to take into account this change.
2. Add no-errors and spec refinement proofs for `mbox_drop`, `mbox_concat_chains`, and `mbox_len`.

NB: The changes to `arrays.v`, `clearbufs,v`, `iter_linked_list.v`, `mbox.v`, `rust_lifetimes.v`, and `string_set.v` are all auto-generated. In a future PR, these auto-generated files are going to be added to the `.gitignore`.

### CODE REVIEW

**Review by eddywestbrook — APPROVED** [review]
The Coq proofs look good! The Haskell code is my own, though, so I'm waiting for someone else (possibly @m-yac himself, who authored this PR) to review.

File: heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs [hunk 1]
```diff
@@ -3079,21 +3102,28 @@ translatePermImpl1 :: ImplTranslateF r ext blocks tops ret =>
...
   ([nuMP| Impl1_Catch |],
    [nuMP| (MbPermImpls_Cons _ (MbPermImpls_Cons _ _ mb_impl1) mb_impl2) |]) ->
-    pitmCatching (translatePermImpl prx $ mbCombine RL.typeCtxProxies mb_impl1) >>= \maybe_trans1 ->
-    pitmCatching (translatePermImpl prx $ mbCombine RL.typeCtxProxies mb_impl2) >>= \maybe_trans2 ->
-    case (maybe_trans1, maybe_trans2) of
-      (Just trans1, Just trans2) ->
+    pitmCatching (translatePermImpl prx $
+                  mbCombine RL.typeCtxProxies mb_impl1) >>= \trans_pair1 ->
+    pitmCatching (translatePermImpl prx $
+                  mbCombine RL.typeCtxProxies mb_impl2) >>= \trans_pair2 ->
```
Comment (m-yac) [inline heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs:3112]: Why not use a pattern matching lambda here (i.e. `... >>= \(mtrans1,hasf1) ->`) like you do below (lies 3150 and 3152)?

**Review by m-yac — COMMENTED** [review]
This all looks good to me. I did comment one readability nitpick, but if you don't care I'd be happy to merge this as-is.

**Comment by eddywestbrook** [comment]
The CI tests all passed before the latest commit, which was just a small tweak and also I tested thoroughly on my local machine, so I am squashing without waiting for the CI. May I be struck down by your angry words if I break the build.

**Pull request merged**